### PR TITLE
Prune imports no longer used by retained entities in the schema

### PR DESCRIPTION
**Pull request by mpeyper**
This is a WIP but wanted to get it up to get some feedback on the approach. This started with a discussion with @swankjesse after I discovered that imports are not necessarily being pruned from files when they are no longer needed by retained entities in the schema.

The existing logic appear to retain an imports when the file is being retained in the schema at all, regarless of whether the import is still being used by the file (and not by another file).

This change attempts to resolve the referenced types in the file and resolve them back to their import locations so that the list of actually required imports can be determined.

It's currently passing all existing test cases as well as the new test we added to replicate the issue, but I'm not sure if I'm capturing all ways a type can be referenced at the moment. Any insights into wierd edge cases that I should consider are very welcome.

I'll add some comments in the files where I'm not sure about things.

### CODE REVIEW

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 1]
```diff
@@ -129,24 +129,37 @@ data class ProtoFile(
     return result
   }
 
+  private fun referencedTypes(schema: Schema): List<Type> {
+    val types = typesAndNestedTypes()
+    val messages = types.filterIsInstance<MessageType>()
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:134]: Is there a better way to get the message types defined in the proto file?

File: wire-schema/src/jvmTest/kotlin/com/squareup/wire/schema/PrunerTest.kt [hunk 2]
```diff
@@ -1462,6 +1462,54 @@ class PrunerTest {
...
+   * even if those imports were unnecessary.
+   */
+
+  @Test
+  fun importPruningIsPrecise() {
+    val schema = buildSchema {
+      add(
+        "message.proto".toPath(),
+        """
+             |import 'footer.proto';
```
Comment (mpeyper) [inline wire-schema/src/jvmTest/kotlin/com/squareup/wire/schema/PrunerTest.kt:1476]: Previously this import was being retained event the `AnotherMessage` was being pruned, along with the need to keep importing it.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 3]
```diff
@@ -129,24 +129,37 @@ data class ProtoFile(
...
+
+    val messageTypes = messages
+      .flatMap { it.fieldsAndOneOfFields }
+      .mapNotNull { it.type }
+
+    val optionTypes = (messages.map { it.options } + messages.flatMap { it.oneOfs }.map { it.options }).flatMap { it.fields().asMap().keys }
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:140]: Note to self, I need to clean this line up and make it more readable.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 4]
```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
...
+    val messageOptions = messages.map { it.options }
+    val fieldOptions = messages.flatMap { it.fieldsAndOneOfFields.map { it.options } }
+    val oneOfOptions = messages.flatMap { it.oneOfs }.map { it.options }
+    val optionTypes = (messageOptions + oneOfOptions + fieldOptions).flatMap { it.fields().asMap().keys }
+
+    val rpcTypes = services
+      .flatMap { service -> service.rpcs }
+      .flatMap { rpc -> listOfNotNull(rpc.requestType, rpc.responseType) }
+
+    return (messageTypes + optionTypes + rpcTypes).distinct().mapNotNull { schema.getType(it) }.toList()
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:149]: * `messageTypes` are the field types defined in messages
* `optionTypes` are the option types for the message and fields, and one-ofs
  * The import paths of any option extensions are resolved below in `retainImports`
* `rpcTypes` are the request and response types defined in services

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 5]
```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
...
-    }
+  fun retainImports(schema: Schema): ProtoFile {
+    val referencedTypes = referencedTypes(schema)
+
+    val typeLocations = referencedTypes.map { type -> type.location }
+
+    val extensionLocations = referencedTypes
+      .filterIsInstance<MessageType>()
+      .flatMap { type -> type.extensionFields }
+      .map { field -> field.location }
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:161]: With the existing test suite, The locations that have been important to retain are the location of a referenced type itself (`typeLocations`), and the location of any option extensions (`extensionLocations`).

Am I missing any other places I should be looking for locations?

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 6]
```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
...
+
+    val referencedImports = (typeLocations + extensionLocations).map { it.path }.toSet()
+
+    val retainedImports = imports.filter { referencedImports.contains(it) }
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:165]: The logic here is that I want to ensure that I only remove existing imports and never add new ones.  I had a case where `google/protobuf/descriptor.proto` was being inserted into the imports if I simply used `referencedImports` as te new `retainedImports`.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 7]
```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
...
 
     return if (imports.size != retainedImports.size) {
       val result = ProtoFile(
```
Comment (mpeyper) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:168]: It's actually on the next line, but I wonder if there is anything that _should_ happen with `publicImports`. Like, if the `public import` is no retained in the schema, perhaps we should prune it out of the file as well?

File: wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/FileDescriptorGeneratorTest.kt [hunk 8]
```diff
@@ -46,7 +46,9 @@ internal class FileDescriptorGeneratorTest {
       |package test;
       |import "imported.proto";
       |
-      |message Test {}
+      |message Test {
+      |  optional Imported imported = 1;
```
Comment (mpeyper) [inline wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/FileDescriptorGeneratorTest.kt:50]: A side effect of this change is that `descriptor.dependencies.size` was 0 without this change becuase the unused import is getting pruned away. I'm not sure how much of an issue that is in reality though?

Comment (oldergod) [reply wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/FileDescriptorGeneratorTest.kt:50]: no problem I think

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt [hunk 9]
```diff
@@ -53,8 +53,9 @@ class Pruner(
   }
 
   private fun retainImports(protoFiles: List<ProtoFile>): List<ProtoFile> {
+    val schema = Schema(protoFiles)
     return protoFiles.map { protoFile ->
-      protoFile.retainImports(protoFiles)
+      protoFile.retainImports(schema)
```
Comment (oldergod) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt:58]: Could we do the transformation of `protoFiles` to `schema` in the `retainImports` method so the public API doesn't change?

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 5 above)
Comment (oldergod) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:161]: I think that's cool

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 4 above)
Comment (oldergod) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:149]: Not sure if we need to retain enum's and constants' options?

**Review by oldergod — APPROVED** [review]
🚀 LGTM

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 4 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:149]: I remembered later that I found a list of a bunch of different option types somewhere. I’ll take a look at that and see if there’s others to add as well.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt (hunk 9 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt:58]: We definitely could. The main concern is we’d be doing the same transformation multiple times for each proto file, but I’m not sure how expensive that is and if it would matter.

Alternatively I could add an overload that takes the schema and have the existing api delegate to it to so the change is only additive instead of breaking?

Comment (oldergod) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt:58]: Yes, ok

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt [hunk 10]
```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
     return result
   }
 
+  private fun referencedTypes(schema: Schema): List<Type> {
```
Comment (swankjesse) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:132]: I think this should return `ProtoType` objects, not `Type` objects. Ie. the call to `mapNotNull()` should be the caller’s responsibility.

(Just to make this more general purpose)

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 6 above)
Comment (swankjesse) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:165]: Good call

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 7 above)
Comment (swankjesse) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:168]: Yeah, good idea

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt (hunk 9 above)
Comment (swankjesse) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt:58]: Thumbs up to adding an overload

**Review by swankjesse — APPROVED** [review]

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 10 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:132]: I went with `Type` for consistency with `typesAndNestedTypes`, but happy to change to `ProtoType` if that’s preferred?

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 4 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt]: I've update this now to look for all of [these option types](https://github.com/square/wire/blob/master/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt#L405-L419) (at least where I believe they would be in the model). Please confirm the new logic makes sense.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 7 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:193]: I've added this, but will need to write some test cases around it.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt (hunk 9 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt:58]: Added and removed the breaking change.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 10 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt]: Done.

Comment (swankjesse) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt]: Thanks.

typesAndNestedTypes() doesn’t need to look things up in a schema because the types are defined within the file.

This thing doesn’t have that information so doing the schema lookup is extra work.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt (hunk 7 above)
Comment (mpeyper) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt:197]: Test added.

**Comment by mpeyper** [comment]
I've added a bunch of tests to test the various ways a type can be referenced and also to test the new `import public` pruning logic. I think I have all cases covered now and I think this is ok to merge now.

**Review by oldergod — APPROVED** [review]
I pushed a commit to satisfy spotless. Very good tests, thank you

**Review by swankjesse — APPROVED** [review]

**Comment by oldergod** [comment]
🚀
